crepecake/src/main/java/com/google/cloud/tools/crepecake/cache/CacheMetadataTranslator.java

```diff
@@ -0,0 +1,108 @@
+/*
+ * Copyright 2017 Google Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not
+ * use this file except in compliance with the License. You may obtain a copy of
+ * the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations under
+ * the License.
+ */
+
+package com.google.cloud.tools.crepecake.cache;
+
+import com.google.cloud.tools.crepecake.blob.BlobDescriptor;
+import com.google.cloud.tools.crepecake.cache.json.CacheMetadataLayerObjectTemplate;
+import com.google.cloud.tools.crepecake.cache.json.CacheMetadataTemplate;
+import com.google.cloud.tools.crepecake.image.DuplicateLayerException;
+import com.google.cloud.tools.crepecake.image.LayerPropertyNotFoundException;
+import java.io.File;
+import java.nio.file.Path;
+
+/** Translates {@link CacheMetadata} to and from {@link CacheMetadataTemplate}. */
+public class CacheMetadataTranslator {
+
+  /** Translates {@link CacheMetadataTemplate} to {@link CacheMetadata}. */
+  static CacheMetadata fromTemplate(CacheMetadataTemplate template, Path cacheDirectory)
+      throws CacheMetadataCorruptedException {
+    try {
+      CacheMetadata cacheMetadata = new CacheMetadata();
+
+      for (CacheMetadataLayerObjectTemplate layerObjectTemplate : template.getLayers()) {
+        File layerContentFile =
+            CacheFiles.getLayerFile(cacheDirectory, layerObjectTemplate.getDigest());
+
+        LayerMetadata layerMetadata;
+        switch (layerObjectTemplate.getType()) {
+          case BASE:
+            layerMetadata =
+                new LayerMetadata(layerObjectTemplate.getType(), layerObjectTemplate.getExistsOn());
+            break;
+
+          case DEPENDENCIES:
+          case RESOURCES:
+          case CLASSES:
+            layerMetadata =
```

Comment:
Why can't it be this unconditionally?

Reply:
Base image layers do not have the fields for `#getSourceDirectories` and `#getLastModifiedTime` and the corresponding methods on `CacheMetadataLayerObjectTemplate` are intended to be only called for application layers (will throw an exception if those properties are not found).

Reply:
I suspected something like that, but in that case I'd expect these fields to return null, which could be passed in here. 

Otherwise, it feels like the class hierarchy is not quite right. Objects shouldn't have methods that don't make sense for some instances of the type. Any method on a class should be applicable to all instances of the type.